diff --git a/drivers/net/wireless/ath/ath6kl/cfg80211.c b/drivers/net/wireless/ath/ath6kl/cfg80211.c
index cce4625..ad4812a 100644
--- a/drivers/net/wireless/ath/ath6kl/cfg80211.c
+++ b/drivers/net/wireless/ath/ath6kl/cfg80211.c
@@ -3779,14 +3779,15 @@ int ath6kl_cfg80211_init(struct ath6kl *ar)
 	 * the firmware has support to override RSN capability, otherwise
 	 * 4-way handshake would fail.
 	 */
-	if (!(ht &&
+	ath6kl_warn("Checking for ATH6KL_FW_CAPABILITY_RSN_CAP_OVERRIDE is disabled.");
+	/*if (!(ht &&
 	      test_bit(ATH6KL_FW_CAPABILITY_RSN_CAP_OVERRIDE,
 		       ar->fw_capabilities))) {
 		ath6kl_band_2ghz.ht_cap.cap = 0;
 		ath6kl_band_2ghz.ht_cap.ht_supported = false;
 		ath6kl_band_5ghz.ht_cap.cap = 0;
 		ath6kl_band_5ghz.ht_cap.ht_supported = false;
-	}
+	}*/
 
 	if (test_bit(ATH6KL_FW_CAPABILITY_64BIT_RATES,
 		     ar->fw_capabilities)) {
