diff -ru linux-3.14/drivers/net/wireless/ath/regd.c linux-3.14.new/drivers/net/wireless/ath/regd.c
--- linux-3.14/drivers/net/wireless/ath/regd.c   2012-03-19 05:15:34.000000000 +0600
+++ linux-3.14.new/drivers/net/wireless/ath/regd.c       2012-06-05 20:47:03.906526443 +0600
@@ -196,6 +196,8 @@
 	struct ieee80211_channel *ch;
 	unsigned int i;
 
+	return;
+
 	for (band = 0; band < IEEE80211_NUM_BANDS; band++) {
 
 		if (!wiphy->bands[band])
@@ -250,6 +252,8 @@
 	struct ieee80211_channel *ch;
 	const struct ieee80211_reg_rule *reg_rule;
 
+	return;
+
 	sband = wiphy->bands[IEEE80211_BAND_2GHZ];
 	if (!sband)
 		return;
@@ -299,6 +303,8 @@
 	struct ieee80211_channel *ch;
 	unsigned int i;
 
+	return;
+
 	if (!wiphy->bands[IEEE80211_BAND_5GHZ])
 		return;
 
@@ -608,6 +614,8 @@
 {
 	const struct ieee80211_regdomain *regd;
 
+	return 0;
+
 	wiphy->reg_notifier = reg_notifier;
 	wiphy->flags |= WIPHY_FLAG_STRICT_REGULATORY;
 
